File: model_compilers/cpp/regression_compiler.cpp
/*
Code converted from Python to C++ using GPT-4o (by OpenAI)

> Original Code is written in Python!

> If you find bugs or have suggestions, please open an issue on GitHub.

Date: 10.05.2025
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cmath>
#include <Eigen/Dense>
#include "json.hpp"

using json = nlohmann::json;
using namespace std;
using namespace Eigen;

vector<string> basis_functions;
vector<double> coefficients;
int n_features;

double apply_basis_function(const VectorXd& x, const string& func) {
    if (func == "1") return 1.0;
    else if (func.rfind("log1p_x", 0) == 0) {
        int idx = stoi(func.substr(8));
        return log1p(abs(x[idx]));
    } else if (func.rfind("exp_x", 0) == 0) {
        int idx = stoi(func.substr(5));
        return exp(min(max(x[idx], -10.0), 10.0));
    } else if (func.rfind("sin_x", 0) == 0) {
        int idx = stoi(func.substr(5));
        return sin(x[idx]);
    } else if (func.find('^') != string::npos) {
        size_t pos = func.find('^');
        int idx = stoi(func.substr(1, pos - 1));
        int power = stoi(func.substr(pos + 1));
        return pow(x[idx], power);
    } else if (func.find(' ') != string::npos) {
        double result = 1.0;
        size_t start = 0, end;
        while ((end = func.find(' ', start)) != string::npos) {
            string token = func.substr(start, end - start);
            result *= x[stoi(token.substr(1))];
            start = end + 1;
        }
        result *= x[stoi(func.substr(start).substr(1))];
        return result;
    } else {
        int idx = stoi(func.substr(1));
        return x[idx];
    }
}

VectorXd predict(const MatrixXd& X) {
    int n_samples = X.rows();
    int n_basis = basis_functions.size();
    MatrixXd X_transformed(n_samples, n_basis);

    for (int i = 0; i < n_samples; ++i) {
        for (int j = 0; j < n_basis; ++j) {
            X_transformed(i, j) = apply_basis_function(X.row(i), basis_functions[j]);
        }
    }

    VectorXd coeffs = Map<VectorXd>(coefficients.data(), coefficients.size());
    return X_transformed * coeffs;
}

int main() {
    // Load model from JSON
    ifstream in("outputs/california_housing_model.json");
    json symbolic_model;
    in >> symbolic_model;
    basis_functions = symbolic_model["basis_functions"].get<vector<string>>();
    coefficients = symbolic_model["coefficients"].get<vector<double>>();
    n_features = symbolic_model["n_features"].get<int>();

    // Generate input data
    MatrixXd X = MatrixXd::Random(10, n_features);
    VectorXd y_pred = predict(X);
    cout << "Predictions:\n" << y_pred << endl;
    return 0;
}

File: model_compilers/python/regression_compiler.py
import json
import numpy as np

def evaluate_basis_functions(X, basis_functions, n_features):
    """
    Evaluates basis functions on the input data.
    
    Parameters:
    -----------
    X : array-like of shape (n_samples, n_features)
        Input data.
    basis_functions : list
        List of basis function strings (e.g., '1', 'x0', 'x0^2', 'x0 x1', 'log1p_x0').
    n_features : int
        Number of input features.
    
    Returns:
    --------
    X_transformed : ndarray of shape (n_samples, n_basis_functions)
        Transformed data matrix.
    """
    X_transformed = np.zeros((X.shape[0], len(basis_functions)))
    for i, func in enumerate(basis_functions):
        if func == '1':
            X_transformed[:, i] = 1
        elif func.startswith('log1p_x'):
            idx = int(func.split('_')[1][1:])
            X_transformed[:, i] = np.log1p(np.abs(X[:, idx]))
        elif func.startswith('exp_x'):
            idx = int(func.split('_')[1][1:])
            X_transformed[:, i] = np.exp(np.clip(X[:, idx], -10, 10))
        elif func.startswith('sin_x'):
            idx = int(func.split('_')[1][1:])
            X_transformed[:, i] = np.sin(X[:, idx])
        elif '^' in func:
            var, power = func.split('^')
            idx = int(var[1:])
            X_transformed[:, i] = X[:, idx] ** int(power)
        elif ' ' in func:
            vars = func.split(' ')
            result = np.ones(X.shape[0])
            for var in vars:
                idx = int(var[1:])
                result *= X[:, idx]
            X_transformed[:, i] = result
        else:
            idx = int(func[1:])
            X_transformed[:, i] = X[:, idx]
    return X_transformed

def predict(X, symbolic_model):
    """
    Predicts target values for the input data.
    
    Parameters:
    -----------
    X : array-like of shape (n_samples, n_features)
        Input data.
    
    Returns:
    --------
    y_pred : ndarray of shape (n_samples,)
        Predicted values.
    """

    X = np.asarray(X)
    X_transformed = evaluate_basis_functions(X, symbolic_model['basis_functions'], 
                                            symbolic_model['n_features'])

    return np.dot(X_transformed, symbolic_model['coefficients'])

if __name__ == "__main__":
    with open('outputs/california_housing_model.json', 'r') as f:
        symbolic_model = json.load(f)
    X = np.random.rand(10, symbolic_model['n_features'])
    y_pred = predict(X, symbolic_model)
    print(y_pred)